analyzer/evmabibackfill/evm_abi_backfill_test.go

```diff
@@ -0,0 +1,298 @@
+package evmabibackfill
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"testing"
+
+	ethCommon "github.com/ethereum/go-ethereum/common"
+
+	"github.com/oasisprotocol/oasis-sdk/client-sdk/go/modules/evm"
+	"github.com/stretchr/testify/require"
+
+	"github.com/oasisprotocol/nexus/analyzer/evmabi"
+	"github.com/oasisprotocol/nexus/common"
+	"github.com/oasisprotocol/nexus/log"
+)
+
+// TODO: Update WROSE source and add tests for more txs, mismatched txs, and custom errors
+
+type mockParsedEvent struct {
+	Name *string
+	Args []*abiEncodedArg
+	Sig  *ethCommon.Hash
+}
+
+type mockParsedTxCall struct {
+	Name *string
+	Args []*abiEncodedArg
+}
+
+func unmarshalEvmEvent(t *testing.T, body []byte) *abiEncodedEvent {
+	var ev evm.Event
+	err := json.Unmarshal(body, &ev)
+	require.Nil(t, err)
+	return &abiEncodedEvent{
+		Round:     0,
+		TxIndex:   nil,
+		EventBody: ev,
+	}
+}
+
+func unmarshalEvmTx(t *testing.T, body string, error_message *string) *abiEncodedTx {
+	txData, err := base64.StdEncoding.DecodeString(body)
+	require.Nil(t, err)
+	return &abiEncodedTx{
+		TxHash:         "", // not relevant for current unit tests
+		TxData:         txData,
+		TxRevertReason: error_message,
+	}
+}
+
+func TestParseEvent(t *testing.T) {
+	abi := evmabi.WROSE
+	p := &processor{
+		runtime: common.RuntimeSapphire,
+		target:  nil,
+		logger:  log.NewDefaultLogger("testing"),
+	}
+	evs := []*abiEncodedEvent{
+		unmarshalEvmEvent(t, []byte(`{"data": "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAADCopscuTr4A=", "topics": ["3fJSrRviyJtpwrBo/DeNqpUrp/FjxKEWKPVaTfUjs+8=", "AAAAAAAAAAAAAAAAw3+DQaxuSpRTgwK81NSc8IUtMMA=", "AAAAAAAAAAAAAAAAWZiaj7/4vj0d6v/ytGtmtjrLeX4="], "address": "i8KwMLKZlk7vteHgs2mRNS5W0tM="}`)),
+		unmarshalEvmEvent(t, []byte(`{"data": "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAxT0Gl9KwCAA=", "topics": ["f89TLBXwptsL1tDgOL6nHTDYCMfZjLO/cmipW/UIG2U=", "AAAAAAAAAAAAAAAAXHL3b32uFt00y2GDtz9HkapLO8Q="], "address": "i8KwMLKZlk7vteHgs2mRNS5W0tM="}`)),
+		unmarshalEvmEvent(t, []byte(`{"data": "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIeGeDJurJAAAA=", "topics": ["4f/8xJI9BLVZ9NKai/xs2gTrWw08RgdRwkAsXFzJEJw=", "AAAAAAAAAAAAAAAA3R7ge0bH7YiGcaZG88ejc5Toz4U="], "address": "i8KwMLKZlk7vteHgs2mRNS5W0tM="}`)),
+		unmarshalEvmEvent(t, []byte(`{"data": "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIeGeDJurJAAAA=", "topics": ["jFvh5evsfVvRT3FCfR6E890DFMD3sikeWyAKyMfDuSU=", "AAAAAAAAAAAAAAAA3R7ge0bH7YiGcaZG88ejc5Toz4U=", "AAAAAAAAAAAAAAAAtTnx0BpDfH8wyvyZTpGPlS3cC6I="], "address": "i8KwMLKZlk7vteHgs2mRNS5W0tM="}`)),
+	}
+
+	name1 := "Transfer"
+	sig1 := ethCommon.HexToHash("0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef")
+	name2 := "Withdrawal"
+	sig2 := ethCommon.HexToHash("0x7fcf532c15f0a6db0bd6d0e038bea71d30d808c7d98cb3bf7268a95bf5081b65")
+	name3 := "Deposit"
+	sig3 := ethCommon.HexToHash("0xe1fffcc4923d04b559f4d29a8bfc6cda04eb5b0d3c460751c2402c5c5cc9109c")
+	name4 := "Approval"
+	sig4 := ethCommon.HexToHash("0x8c5be1e5ebec7d5bd14f71427d1e84f3dd0314c0f7b2291e5b200ac8c7c3b925")
```

Comment:
These are not reused anywhere, so let's inline them (with `common.Ptr()`, if needed).

In general, the rule of thumb with unit test is to keep control flows as simple as possible, inline things as much as possible, reduce the number of helper funcs, etc, all with the goal of making the test linearly readable. Tests of that kind are easier to fix and (IMO even more importantly) read, so they can serve as pretty good documentation on how a function is expected to be used, without having to wade through 100 abstraction layers in the actual non-test program.